robot_skills/src/robot_skills/robot.py

```diff
@@ -126,34 +126,75 @@ def tf_transform_pose(self, ps, frame):
         output_pose = self.tf_listener.transformPose(frame, ps)
         return output_pose
 
-    def get_arm(self, side):
-        """Get an arm object and a backup for that arm by giving a side as either a string or an Arm-object
-        @param side Either string from robot.arms.keys() or Arm from robot.arms.values()
-        >>> robot = Robot("dummy")
-        >>> arm, backup_arm = robot.get_arm("left")
-        >>> assert(arm == robot.leftArm)
-        >>> assert(backup_arm == robot.rightArm)"""
-        preferred_side = self.arms[self.arms.keys()[0]]
-
-        #Define which arm is which's backup arm (left backs up for right etc)
-        backup_arms = self.arms.values() #Get a *list* of arms i.e. the values of the arm-dict, not the keys
-        backup_arms.insert(0, backup_arms.pop()) #Make the last arm the first in the list, so we shift by 1
-        backup_str_dict = dict(zip(self.arms.keys(), backup_arms)) #Create a dict again that maps strings to backup-arms
-        backup_obj_dict = {self.arms[side]:backup_str_dict[side] for side in self.arms.keys()} #Create a dict that maps e.g. self.LeftArm to self.rightArm
-
-        if isinstance(side, basestring):
-            try:
-                preferred_side = self.arms[side]
-            except KeyError:
-                print "Unknown arm side:" + str(side) + ". Defaulting to 'right'"
-                preferred_side = self.arms[self.arms.keys()[0]]
-        elif isinstance(side, arms.Arm):
-            preferred_side = side
-        else:
-            print "Unknown arm side:" + str(side) + ". Defaulting to '{0}'".format(preferred_side.side)
+    def get_robot_arm(self, gripper_type=None, goals=None, trajectories=None, preferred=None):
+        """
+        Find an arm that has the needed properties.
+
+        :param gripper_type: Kind of gripper, currently only one type supported, use 'yes' to select it.
+        :type  gripper_type: string or None
+
+        :param goals: Collection of joint goals that the arm must be able to perform.
+        :type  goals: container wth strings or None.
+
+        :param trajectories: Collection of joint trajectories that the arm must be able to perform.
+        :type  trajectories: container wth strings or None.
+
+        :param preferred_name: Name of the preferred arm if set. By default, this is a soft
+                requirement (as in, you may get a different arm instead if the preferred arm
+                is not suitable). By appending an exclamation mark to the name, it becomes a
+                hard requirement (you will either get the preferred arm or no arm at all).
+        :type  preferred_name: Name (string) of the preferred arm, or None if any arm will do.
```

Comment:
I'd like to see this difference make explicit, not based on string-matching. So, for example only give 'left' or 'right' in preferred arm, but if you really really need a specific arm, then use eg. the ```required``` parameter.

Reply:
All properties can be queried using the `required_xyz` parameter in which case the request must be met. Other 'nice to have' properties can be requested using the `desired_xyz` parameter.

If you get an arm, you can query if a certain property value has been granted.

Currently you get the first arm with all the required properties, which may not be the best choice. In the future, it may be better to try the other arms too, and try to get the largest possible subset of properties (for some yet unknown notion of 'largest').